File: ts/src/satisfy.ts
import type { ConfigDoc, Element, ElementCap, Issue, Manifest, Node, Screen, Source } from "./types";
import { expand, countTiles, depth } from "./presets";

function resolveLayout(screen: Screen, className: string): Node {
  const v = screen.variants?.find((x) => x.class === className);
  return v ? v.layout : screen.layout;
}

function elementIds(n: Node, out: string[]): void {
  if ("element" in n) out.push(n.element);
  else if ("children" in n) n.children.forEach((c) => elementIds(c, out));
  else if ("cells" in n) n.cells.forEach((c) => elementIds(c, out));
}

function checkElement(
  el: Element,
  path: string,
  allowedTypes: Set<string>,
  capByType: Map<string, ElementCap>,
  allowedSources: Set<string>,
  allowedActions: Set<string>,
  issues: Issue[],
): void {
  if (!allowedTypes.has(el.type)) {
    issues.push({ path: `${path}/type`, message: `element type not supported: ${el.type}` });
    return;
  }
  const cap = capByType.get(el.type);
  if (el.bindings) {
    for (const [field, src] of Object.entries(el.bindings)) {
      if (cap?.bindings && !cap.bindings.includes(field))
        issues.push({ path: `${path}/bindings/${field}`, message: `binding field not supported by ${el.type}: ${field}` });
      if (!allowedSources.has((src as Source).kind))
        issues.push({ path: `${path}/bindings/${field}`, message: `source kind not supported: ${(src as Source).kind}` });
    }
  }
  if (el.action && allowedActions.size > 0 && !allowedActions.has(el.action.kind))
    issues.push({ path: `${path}/action`, message: `action kind not supported: ${el.action.kind}` });
}

export function satisfies(config: ConfigDoc, manifest: Manifest, className: string): Issue[] {
  const issues: Issue[] = [];
  const cls = manifest.classes.find((c) => c.id === className);
  if (!cls) return [{ path: "/", message: `class not supported: ${className}` }];

  const allowedTypes = new Set(cls.elements ?? manifest.elements.map((e) => e.type));
  const capByType = new Map(manifest.elements.map((e) => [e.type, e]));
  const allowedSources = new Set(manifest.sources ?? ["signalk"]);
  const allowedActions = new Set(manifest.actionKinds ?? []);

  config.screens.forEach((screen, si) => {
    let tree: Node;
    try {
      tree = expand(resolveLayout(screen, className));
    } catch (e) {
      issues.push({ path: `/screens/${si}/layout`, message: (e as Error).message });
      return;
    }
    const tiles = countTiles(tree);
    if (tiles > cls.maxTiles)
      issues.push({ path: `/screens/${si}/layout`, message: `too many tiles: ${tiles} > ${cls.maxTiles}` });
    const d = depth(tree);
    if (d > cls.maxDepth)
      issues.push({ path: `/screens/${si}/layout`, message: `nesting too deep: ${d} > ${cls.maxDepth}` });

    const ids: string[] = [];
    elementIds(tree, ids);
    for (const id of ids) {
      const el = screen.elements[id];
      if (!el) {
        issues.push({ path: `/screens/${si}`, message: `layout references unknown element: ${id}` });
        continue;
      }
      checkElement(el, `/screens/${si}/elements/${id}`, allowedTypes, capByType, allowedSources, allowedActions, issues);
    }
  });
  return issues;
}

File: py/src/yey_boats_midl/satisfy.py
# SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
# Copyright (c) 2026 Yey Boats Project. See LICENSE and COMMERCIAL.md.
"""Capability satisfaction against a target class — mirrors ts/src/satisfy.ts."""
from __future__ import annotations

from typing import Any, Dict, List, Set

from .presets import count_tiles, depth, expand
from .types import Issue


def _resolve_layout(screen: Dict[str, Any], class_name: str) -> Any:
    for v in screen.get("variants") or []:
        if v.get("class") == class_name:
            return v["layout"]
    return screen.get("layout")


def _element_ids(n: Any, out: List[str]) -> None:
    if "element" in n:
        out.append(n["element"])
    elif "children" in n:
        for c in n["children"]:
            _element_ids(c, out)
    elif "cells" in n:
        for c in n["cells"]:
            _element_ids(c, out)


def _check_element(
    el: Dict[str, Any],
    path: str,
    allowed_types: Set[str],
    cap_by_type: Dict[str, Dict[str, Any]],
    allowed_sources: Set[str],
    allowed_actions: Set[str],
    issues: List[Issue],
) -> None:
    if el["type"] not in allowed_types:
        issues.append(Issue(f"{path}/type", f"element type not supported: {el['type']}"))
        return
    cap = cap_by_type.get(el["type"])
    bindings = el.get("bindings")
    if bindings:
        for fieldname, src in bindings.items():
            if cap and cap.get("bindings") is not None and fieldname not in cap["bindings"]:
                issues.append(
                    Issue(f"{path}/bindings/{fieldname}", f"binding field not supported by {el['type']}: {fieldname}")
                )
            if src.get("kind") not in allowed_sources:
                issues.append(
                    Issue(f"{path}/bindings/{fieldname}", f"source kind not supported: {src.get('kind')}")
                )
    action = el.get("action")
    if action and len(allowed_actions) > 0 and action.get("kind") not in allowed_actions:
        issues.append(Issue(f"{path}/action", f"action kind not supported: {action.get('kind')}"))


def satisfies(config: Dict[str, Any], manifest: Dict[str, Any], class_name: str) -> List[Issue]:
    issues: List[Issue] = []
    cls = next((c for c in manifest.get("classes", []) if c.get("id") == class_name), None)
    if cls is None:
        return [Issue("/", f"class not supported: {class_name}")]

    allowed_types: Set[str] = set(cls.get("elements") or [e["type"] for e in manifest["elements"]])
    cap_by_type: Dict[str, Dict[str, Any]] = {e["type"]: e for e in manifest["elements"]}
    allowed_sources: Set[str] = set(manifest.get("sources") or ["signalk"])
    allowed_actions: Set[str] = set(manifest.get("actionKinds") or [])

    for si, screen in enumerate(config["screens"]):
        try:
            tree = expand(_resolve_layout(screen, class_name))
        except Exception as e:  # noqa: BLE001 — mirror TS catch
            issues.append(Issue(f"/screens/{si}/layout", str(e)))
            continue

        tiles = count_tiles(tree)
        if tiles > cls["maxTiles"]:
            issues.append(Issue(f"/screens/{si}/layout", f"too many tiles: {tiles} > {cls['maxTiles']}"))
        d = depth(tree)
        if d > cls["maxDepth"]:
            issues.append(Issue(f"/screens/{si}/layout", f"nesting too deep: {d} > {cls['maxDepth']}"))

        ids: List[str] = []
        _element_ids(tree, ids)
        for el_id in ids:
            el = (screen.get("elements") or {}).get(el_id)
            if not el:
                issues.append(Issue(f"/screens/{si}", f"layout references unknown element: {el_id}"))
                continue
            _check_element(el, f"/screens/{si}/elements/{el_id}", allowed_types, cap_by_type, allowed_sources, allowed_actions, issues)

    return issues
